tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -2194,14 +2194,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
     dataset = self._apply_options()
     if shared_name is None:
       shared_name = ""
-    iterator_resource = gen_dataset_ops.iterator_v2(
+
+    with ops.colocate_with(self._variant_tensor):
```

Comment:
In my experience colocate_with is very dangerous, and there's almost always a better way of achieving what we want than using colocate_with.

One problem is that `colocate_with(a): b` has different behavior in eager and graph modes: in eager mode it can only affect the placement of b, but in graph mode it also "retroactively" affects the placement of a as well. This is made doubly bad by the fact that colocation group sizes tend to grow out of control quickly (b is rarely a single op, it's often a bunch of ops, many with no particular reason to be there, and all these ops are now tied to the same colocation group, and any device placement constraints or kernel registration issues affecting one of them now affect all of them) and can have very weird behavior as well on multi-node clusters.

Given this I strongly prefer that we:
 1. ask users who are creating their own iterators from datasets to explicitly add device scopes to get the iterators on the right device; explicit here is much less buggy than implicit.
 2. change the distribution strategies which work on GPUs (like MirroredStrategy and the multi-worker equivalent) to add these device scopes, since at that point we know what the right user intent is
 3. figure out a way for horovod to signal user intent in a similar way as distribution strategy (or even implement a horovod strategy, for uniformity with the rest of TF, full keras compatibility, etc); it'll make it less magic.

In the long run I agree it'd be nice if we could manage to not lose information about the dataset when creating iterators; I don't know whether we want to propagate some metadata from python or if we want to propagate some metadata within the runtime and make it visible to the placer. Since this is unclear I'd prefer to defer this decision while unblocking people by supporting explicit placement.

Does that sound reasonable? I'm happy to jump into a videoconference call to clarify my position / listen to counter arguments.

Reply:
@alextp thanks you so much for you answer :) I wasn't aware of this issue with `colocate_with`.

I did something really naive and replaced `with ops.colocate_with()` to `with ops.device()`.
That actually perfectly works and even pass my unittest easy.

```bash
bazel test --verbose_failures //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test
```

Could this be acceptable ?

it has the advantage to completely fix the colocation issue you raise and fully address point 1) and 2) without any additional change. And most likely will be good for Horovod (I gotta try).